src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+     * @throws Exception
+     */
+    protected void downloadData() throws Exception {
+
+        // TODO: Add unit tests
+        // Get names of all organisms we add links and/or microarray data for.
+        // Species names are in biomart format (eg: hsapiens).
+        for (String biomartSpeciesName : getBiomartSpeciesNames()) {
+            logger.info("Retrieving Biomart files for " + biomartSpeciesName);
+            // Create directory where Biomart files will be stored.
+            new File(this.destination).mkdirs();
+
+            logger.info("Retrieving microarray data");
+            // Query Biomart for existing microarray 'types' (not ids) that exist for this species.
+            Set<String> microarrayTypes = queryBiomart(getMicroarrayTypesQuery(biomartSpeciesName), 0);
+            // Iterate through each microarray type and retrieve Ensembl-Microarray identifier mappings.
+            // All mappings are stored in a single file, (eg: hsapiens_microarray);
+            retrieveAndStoreMicroarrayData(biomartSpeciesName, getBiomartXMLPath(), microarrayTypes);
+
+            // Query Ensembl-Uniprot (swissprot and trembl) identifier mapping data from Biomart
+            // and write it to a file (eg: hsapiens_uniprot).
+            logger.info("Retrieving UniProt data");
+            retrieveAndStoreUniprotData(biomartSpeciesName, getBiomartXMLPath());
+
+            logger.info("Completed Biomart data retrieval for " + biomartSpeciesName);
+        }
+    }
+
+    /**
+     * Query Biomart for both Uniprot mapping types (curated -- SwissProt, generated -- TrEMBL).
+     * @param biomartSpeciesName
+     * @throws Exception
+     */
+    private void retrieveAndStoreUniprotData(String biomartSpeciesName, String biomartXMLPath) throws IOException {
+        Set<String> uniprotIdLines = new HashSet<>();
+        try {
+            for (String uniprotQueryId : Arrays.asList("uniprotswissprot", "uniprotsptrembl")) {
+                // If proper response, write data to file.
+                Set<String> uniprotResponseLines = queryBiomart(getBiomartIdentifierQuery(biomartXMLPath, biomartSpeciesName, uniprotQueryId), 0);
+                for (String uniprotResponseLine : uniprotResponseLines) {
+                    List<String> tabSplit = Arrays.asList(uniprotResponseLine.split("\t"));
+                    if (tabSplit.size() == 4) {
+                        uniprotIdLines.add(uniprotResponseLine + "\n");
+                    }
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        // Write data to file.
+        if (!uniprotIdLines.isEmpty()) {
+            File speciesUniprotFile = createNewFile(this.destination + biomartSpeciesName + "_uniprot");
+            for (String uniprotIdLine : uniprotIdLines) {
+                Files.write(Paths.get(speciesUniprotFile.toString()), uniprotIdLine.getBytes(), StandardOpenOption.APPEND);
+            }
```

Comment:
This could be written as:
```
for (String uniprotIdLine : uniprotIdLines) {
    Files.write(Paths.get(this.destination, biomartSpeciesName + "_uniprot", uniprotIdLine.getBytes(), StandardOpenOption.APPEND, StandardOpenOption.CREATE);
}
```

This will create the file if it doesn't exist and append to it if it does.  That way checking for the empty set isn't needed since the for loop won't be entered if the set is empty and no file will be created.

Reply:
Thats neat, I didn't realize you could do that with the Append/Create

Reply:
Yup, you can specify unlimited "OpenOption" values after the first two parameters and combining them has different effects:
https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#write-java.nio.file.Path-java.lang.Iterable-java.nio.file.OpenOption...-